substrate/frame/transaction-payment/skip-feeless-payment/src/lib.rs

```diff
@@ -0,0 +1,145 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//! # Skip Feeless Payment Pallet
+//!
+//! This pallet allows runtimes that include it to skip payment of transaction fees for
+//! dispatchables marked by [`#[pallet::feeless_if]`](`macro@
+//! frame_support::pallet_prelude::feeless_if`).
+//!
+//! ## Overview
+//!
+//! It does this by wrapping an existing [`SignedExtension`] implementation (e.g.
+//! [`pallet-transaction-payment`]) and checking if the dispatchable is feeless before applying the
+//! wrapped extension. If the dispatchable is indeed feeless, the extension is skipped and a custom
+//! event is emitted instead. Otherwise, the extension is applied as usual.
+//!
+//!
+//! ## Integration
+//!
+//! This pallet wraps an existing transaction payment pallet. This means you should both pallets
+//! in your `construct_runtime` macro and include this pallet's
+//! [`SignedExtension`] ([`SkipCheckIfFeeless`]) that would accept the existing one as an argument.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::{Decode, Encode};
+use frame_support::{
+	dispatch::{CheckIfFeeless, DispatchResult},
+	traits::{IsType, OriginTrait},
+};
+use scale_info::TypeInfo;
+use sp_runtime::{
+	traits::{DispatchInfoOf, PostDispatchInfoOf, SignedExtension},
+	transaction_validity::TransactionValidityError,
+};
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A transaction fee was skipped.
+		FeeSkipped { who: T::AccountId },
```

Comment:
Was it intentional that this is a pallet? as in, is this event really needed, to warrant the existence of a pallet? 

If coded a bit differently, we could make it possible for you to use `SkipCheckIfFeeless` without needing the pallet in the runtime. 

The fact that now, fee-related events are scattered into events coming from two different pallets is IMO ever more reason to bake fees into `frame-system` once and for all.

Reply:
Yeah, the intent was to entirely skip a call to the payment extension (be it the ones provided in this repo or a custom one) if a call is feeless, but still log that as an event. What could be a possible alternative?

> ever more reason to bake fees into frame-system once and for all
Happy to brainstorm on this. Is there an issue where this is described already? Or should I go ahead and create one?